### ISSUE TITLE
Json.NET types not serialized properly

### ISSUE THREAD

**Issue by tmenier** [elastic/elasticsearch-net#3228]
Objects of any type deriving from [JToken](https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_Linq_JToken.htm) are not being serialized with their values. All values are serialized as empty arrays. For example, an object with properties of types `JValue`, `JArray`, `JObject` fully populated with non-null values, will get serialized like this:

```
{
    "myJValue": [],
    "myJArray": [
        [],
        [],
        []
    ],
    "myJObject": {
        "a": [],
        "b": [],
        "c": {
            "x": [],
            "y": []
        }
    }
}
```

I posted a related [question](https://stackoverflow.com/questions/50107494/how-do-i-serialize-properties-of-type-jtoken-or-jobject-in-elasticsearch-nest) on Stack Overflow. Is this a bug? I'm aware you're using Json.NET under the hood and out of the box I know it doesn't have trouble serializing these types, so I'm wondering if it's something to do with your custom default serializer.

**Comment by russcam** [comment]
Hey @tmenier, I answered your question on Stackoverflow; let me know there if you need any further details 👍 I'm going to close this issue.

**Comment by tmenier** [comment]
Thanks. With respect this being "expected behavior" (as you noted on SO), I think that depends a little on perspective. Regardless of any internal implementation details, when the values some dictionary-like thing (or object-like thing) are replaced by empty arrays, it's hard to see how the average user would expect that. I know it can't possibly know about every 3rd-party type that might be considered "dictionary-like", and it's probably just falling back on default behavior here, but an argument _could_ be made that `JToken`/`JObject` are so ubiquitous in the .NET/JSON world that maybe they deserve special treatment (without installing a different package).

**Comment by Mpdreamz** [comment]
`JToken` implements `IEnumerable<>` though, so its not really a dictionary/object like thing as we'd might think of `JToken`.  `JObject` does implement `IDictionary<string, JToken>` so we do see the keys in the output as expected. So as far as known types it is behaving as expected.

One *could* argue that they are ubiquitous but unless these types ship as part of the BCL or default netstandard library I personally don't view these as something we need to support OOTB. We do support it however via an additional package which I think is the best default our users. It might not be your best default but for someone who would like to use JIL for their source serialization pulling in `JSON.NET` is not a good default either. The current setup serves both ends of the spectrum here.

Additional reasons for no longer depending on `Json.NET` by default are:

* Your `JSON.NET` dependency is no longer tied to the whole client, you can now simply pick the `NEST.JsonNetSerializer` version that matches the `JSON.NET` dep you need.
* Your `JSON.NET` configuration does not bleed into ours anymore when you pass a custom serializer.
* We are free to internally switch away from JSON.NET and prebake all of our serialization routines.

**Comment by tmenier** [comment]
Points well taken. "Knowing" about specific types from libraries it doesn't explicitly depend on does seem like something that shouldn't be expected, and exactly why `NEST.JsonNetSerializer` exists.

Regarding the interfaces that decedents of  `JToken` implement, (`IEnumerable`, `IDictionary`, etc.) NEST _almost_ gets there without specific knowledge of Json.NET types. `JValue` might be the one specific case causing to to fall short. I'm reminded of something a little unorthodox I did in Flurl to [try and infer if something is "dictionary-like" by looking for common property names](https://github.com/tmenier/Flurl/blob/master/src/Flurl/Util/CommonExtensions.cs#L87-L102) (`Key` or  `Name`, `Value`). [`JObject` has a `Value` property](https://www.newtonsoft.com/json/help/html/P_Newtonsoft_Json_Linq_JValue_Value.htm) so something similar could work here. But again, what I did is a bit opinionated - what I call making the best attempt possible to infer intent, others might call a bad hack. :)
